server/src/main/java/org/apache/iotdb/db/query/expression/unary/ConstantExpression.java

```diff
@@ -0,0 +1,111 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.query.expression.unary;
+
+import org.apache.iotdb.db.exception.query.LogicalOptimizeException;
+import org.apache.iotdb.db.exception.query.QueryProcessException;
+import org.apache.iotdb.db.metadata.PartialPath;
+import org.apache.iotdb.db.qp.physical.crud.UDTFPlan;
+import org.apache.iotdb.db.qp.utils.WildcardsRemover;
+import org.apache.iotdb.db.query.expression.Expression;
+import org.apache.iotdb.db.query.udf.core.executor.UDTFExecutor;
+import org.apache.iotdb.db.query.udf.core.layer.ConstantIntermediateLayer;
+import org.apache.iotdb.db.query.udf.core.layer.IntermediateLayer;
+import org.apache.iotdb.db.query.udf.core.layer.LayerMemoryAssigner;
+import org.apache.iotdb.db.query.udf.core.layer.RawQueryInputLayer;
+import org.apache.iotdb.db.utils.CommonUtils;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+
+import java.time.ZoneId;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/** Constant expression */
+public class ConstantExpression extends Expression {
+
+  private final Object value;
+  private final TSDataType dataType;
+
+  public ConstantExpression(TSDataType dataType, String str) throws QueryProcessException {
+    this.dataType = dataType;
+    this.value = CommonUtils.parseValue(dataType, str);
+  }
+
+  public Object getValue() {
+    return value;
+  }
+
+  public TSDataType getDataType() {
+    return dataType;
+  }
+
+  @Override
+  public boolean isPureConstantExpression() {
+    return true;
+  }
+
+  @Override
+  public void concat(List<PartialPath> prefixPaths, List<Expression> resultExpressions) {
+    resultExpressions.add(this);
+  }
+
+  @Override
+  public void removeWildcards(WildcardsRemover wildcardsRemover, List<Expression> resultExpressions)
+      throws LogicalOptimizeException {
+    resultExpressions.add(this);
+  }
+
+  @Override
+  public void collectPaths(Set<PartialPath> pathSet) {
+    // Do nothing
+  }
+
+  @Override
+  public void constructUdfExecutors(
+      Map<String, UDTFExecutor> expressionName2Executor, ZoneId zoneId) {
+    // Do nothing
+  }
+
+  @Override
+  public void updateStatisticsForMemoryAssigner(LayerMemoryAssigner memoryAssigner) {
+    // Do nothing
+  }
+
+  @Override
+  public IntermediateLayer constructIntermediateLayer(
+      long queryId,
+      UDTFPlan udtfPlan,
+      RawQueryInputLayer rawTimeSeriesInputLayer,
+      Map<Expression, IntermediateLayer> expressionIntermediateLayerMap,
+      Map<Expression, TSDataType> expressionDataTypeMap,
+      LayerMemoryAssigner memoryAssigner) {
+    expressionDataTypeMap.put(this, this.getDataType());
+    IntermediateLayer intermediateLayer =
+        new ConstantIntermediateLayer(this, queryId, memoryAssigner.assign());
+    expressionIntermediateLayerMap.put(this, intermediateLayer);
+    return intermediateLayer;
```

Comment:
For 

select sin(1 + s1), cos(1 + s1) from root.sg.d

we'd better not construct the IntermediateLayer twice.

Reply:
Will cache the intermediate layer for the same const expression.